File: tikzer/tikzer.cpp
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>
#include <unistd.h>
#include <string>

using namespace std;

int n;
struct Node
{
    int id;
    char sis, aua;
};

int idd;
char ss[20], aa[20];

Node nodes[1001];
double mat[1001][1001];

inline void gen_tikz()
{
    FILE *g = fopen("ret.tex", "w");
    fprintf(g, "\\documentclass[crop,tikz]{standalone}\n");
    fprintf(g, "\\definecolor{mygreen}{HTML}{006400}\n");
    fprintf(g, "\\definecolor{myred}{HTML}{8B0000}\n");
    fprintf(g, "\\begin{document}\n");
    fprintf(g, "\\begin{tikzpicture}[transform shape,line width=2pt]\n");
    double R = 1.0;
    while (2.0 * R * sin(M_PI/n) < 3) R += 0.1;
    for (int i=0;i<n;i++)
    {
        double pos = (i * 360.0) / (n * 1.0);
        string color = "white";
        if (nodes[i].sis == 'S')
        {
            color = "blue";
        } 
        if (nodes[i].sis == 'I')
        {
            color = "myred";
        }
        else if (nodes[i].sis == 'V')
        {
            color = "mygreen";
        }
        fprintf(g, "\\node[draw,circle,inner sep=0.25cm,fill=%s,text=white] (N-%d) at (%lf:%lfcm) [thick] {$%c$};\n", color.c_str(), i, pos, R, nodes[i].aua);
    }
    for (int i=0;i<n;i++)
    {
        for (int j=0;j<i;j++)
        {
            double w = fmax(0.01, mat[i][j]);
            fprintf(g, "\\path(N-%d) edge[-, red, opacity=%lf] (N-%d);\n", i, w, j);
        }
    }
    fprintf(g, "\\end{tikzpicture}\n");
    fprintf(g, "\\end{document}\n");
    fclose(g);
}

inline void build_tikz()
{
    system("exec pdflatex ret.tex &> /dev/null");
    system("exec convert -density 300 ret.pdf -quality 90 ret.png &> /dev/null");
}

int main()
{
    while (true)
    {
        FILE *f = fopen("final_log.log", "r");
        if (f == NULL)
        {
            usleep(5000000);
            continue;
        }
        fscanf(f, "%*s");
        fscanf(f, "%d%*d", &n);
        for (int i=0;i<n;i++)
        {
            fscanf(f, "%d%s%s", &idd, aa, ss);
            nodes[i].id = idd;
            nodes[i].sis = ss[0];
            nodes[i].aua = aa[0];
        }
        for (int i=0;i<n;i++)
        {
            double sum = 0.0;
            for (int j=0;j<n;j++)
            {
                fscanf(f, "%lf", &mat[i][j]);
                sum += mat[i][j];
            }
            if (sum > 0.0)
            {
                for (int j=0;j<n;j++)
                {
                    mat[i][j] /= sum;
                }
            }
        }
        printf("Data processed.\n");
        gen_tikz();
        printf("TikZ generated.\n");
        build_tikz();
        printf("TikZ built.\n");
        fclose(f);
        usleep(5000000);
    }
    return 0;
}

File: server/src/main/java/com/hackbridge/viral/Tikzer.java
package com.hackbridge.viral;

import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpHandler;
import com.sun.net.httpserver.HttpServer;

import java.io.*;
import java.net.InetSocketAddress;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Map;

/**
 * The TikZ visualisation server (Tikzer).
 * It continuously generates TikZ figures from the most recent state log, and HTTP serves the
 * most recently generated figure to the client; listening on port 8000 at /viz.
 */
class Tikzer {
    private static int maxLogs = 100; // this should be enough to avoid any issues with loading the image
    private HttpServer tikzer;
    private StateLog latestLog;
    private String mostRecentLogFile;
    private int currentId;

    private final Object guard = new Object();
    private boolean hasChanged = false;

    private static String base = "ret_";

    Tikzer(int port) {
        try {
            currentId = 0;
            mostRecentLogFile = "";
            tikzer = HttpServer.create(new InetSocketAddress(port), 0);
            tikzer.createContext("/viz", new VizHandler(this));
            tikzer.setExecutor(null); // creates a default executor
            tikzer.start();
            this.launch();
        } catch (IOException ioe) {
            ioe.printStackTrace();
        }
    }

    private void launch() {
        Thread t = new Thread() {
            @Override
            public void run() {
                while (true) {
                    StateLog curr;
                    synchronized (guard) {
                        while (!hasChanged) {
                            try {
                                guard.wait();
                            } catch (InterruptedException e) {
                                e.printStackTrace();
                            }
                        }
                        curr = latestLog;
                        hasChanged = false;
                    }
                    try {
                        int nextId = (currentId + 1) % maxLogs;
                        mostRecentLogFile = TikzWorker.generate(curr, nextId);
                        currentId = nextId;
                    } catch (IOException ioe) {
                        ioe.printStackTrace();
                    }
                }
            }
        };
        t.setDaemon(true);
        t.start();
    }

    void addLog(StateLog s) {
        synchronized (guard) {
            latestLog = s;
            hasChanged = true;
            guard.notify();
        }
    }

    private static class TikzWorker {
        static void gen_tikz(StateLog log, String tikzName) throws IOException {
            Map<Long, Node> nodes = log.getNodes();
            Map<Long, Integer> positions = log.getPositionMap();
            List<ArrayList<Double>> state = log.getState();

            BufferedWriter tikzFile = new BufferedWriter(new FileWriter(tikzName));
            tikzFile.write("\\documentclass[crop,tikz]{standalone}\n");
            tikzFile.write("\\definecolor{mygreen}{HTML}{006400}\n");
            tikzFile.write("\\definecolor{myred}{HTML}{8B0000}\\n");
            tikzFile.write("\\begin{document}\n");
            tikzFile.write("\\begin{tikzpicture}[transform shape,line width=2pt]\n");

            double R = 1.0;
            int n = nodes.size();
            int i = 0;
            while (2.0 * R * Math.sin(Math.PI / n) < 3) R += 0.1;
            for (Map.Entry<Long, Node> entry : nodes.entrySet()) {
                Long index = entry.getKey();
                Node currNode = entry.getValue();
                double pos = (i * 360.0) / n;
                String color = "white";
                switch (currNode.getPhysicalState()) {
                    case SUSCEPTIBLE: color = "blue"; break;
                    case INFECTED:
                    case CARRIER: color = "myred"; break;
                    case VACCINATED: color = "mygreen"; break;
                }
                char awareness = '\0';
                switch (currNode.getAwarenessState()) {
                    case AWARE: awareness = 'A'; break;
                    case UNAWARE: awareness = 'U'; break;
                }
                tikzFile.write(String.format(
                        "\\node[draw,circle,inner sep=0.25cm,fill=%s,text=white] (N-%d) at (%f:%fcm) [thick] {$%c$};\n",
                        color, index, pos, R, awareness));
                i++;
            }

            HashSet<Long> used = new HashSet<Long>();

            for (Map.Entry<Long, Integer> i_entry : positions.entrySet()) {
                Long i_id = i_entry.getKey();
                int i_pos = i_entry.getValue();
                used.add(i_id);
                for (Map.Entry<Long, Integer> j_entry : positions.entrySet()) {
                    Long j_id = j_entry.getKey();
                    int j_pos = j_entry.getValue();
                    if (used.contains(j_id)) continue;
                    double w = Math.max(0.01, state.get(i_pos).get(j_pos));
                    tikzFile.write(String.format(
                            "\\path(N-%d) edge[-, red, opacity=%f] (N-%d);\n",
                            i_id, w, j_id));
                }
            }

            tikzFile.write("\\end{tikzpicture}\n");
            tikzFile.write("\\end{document}\n");

            tikzFile.close();
        }

        static void build_tikz(String tikzName, String pdfName, String pngName) throws IOException {
            Runtime.getRuntime().exec("exec pdflatex " + tikzName + " &> /dev/null");
            Runtime.getRuntime().exec("exec convert -density 300 " + pdfName + " -quality 90 " + pngName + " &> /dev/null");
        }

        static String generate(StateLog log, int id) throws IOException {
            String tikzName = base + id + ".tex";
            String pdfName = base + id + ".pdf";
            String pngName = base + id + ".png";
            gen_tikz(log, tikzName);
            build_tikz(tikzName, pdfName, pngName);
            return pngName;
        }
    }

    private static class VizHandler implements HttpHandler {
        Tikzer tkz;

        VizHandler(Tikzer tkz) {
            this.tkz = tkz; // this is by reference, so always store a pointer to its parent
        }

        public void handle(HttpExchange t) throws IOException {
            File file = new File(tkz.mostRecentLogFile);
            t.sendResponseHeaders(200, file.length());
            OutputStream outputStream = t.getResponseBody();
            Files.copy(file.toPath(), outputStream);
            outputStream.close();
        }
    }
}
